### DISCUSSION TITLE
Further information on package loading

### DISCUSSION THREAD

**Discussion by jl5000** [ropensci-books/targets#81] [Help]
As I use targets, specifically reading the documentation around using packages, there are a couple of questions I have which I feel would be useful to address somewhere:

1. The typical way I write functions is to specify the namespace for each function call. This can effectively render the `packages` argument redundant when defining the pipeline. Is there any guidance for the best approach to take?
2. If we use the `packages` argument for multiple targets with the same package(s), could the package(s) be loaded more than once? Or does it just get loaded when it's first needed and then stays loaded?
3. Does the `packages` argument need to specify ALL packages required for the target, or only those in addition to those set globally in `tar_option_set()`?

**Comment by wlandau** [comment]
> The typical way I write functions is to specify the namespace for each function call. This can effectively render the packages argument redundant when defining the pipeline. Is there any guidance for the best approach to take?

Preferences vary from person to person and project to project. `targets` simply makes both options possible. As for which is better, this is a discussion around general R usage and is outside the scope of `targets`. 

> If we use the packages argument for multiple targets with the same package(s), could the package(s) be loaded more than once? Or does it just get loaded when it's first needed and then stays loaded?

`library()` is always called on all the packages before a target runs. If the package is already loaded, then `library()` has no effect. As for whether a package is already loaded, this depends on your [high-performance computing](https://books.ropensci.org/targets/crew.html) configuration. If you don't use high-performance computing, then all the targets will run in the same R session, so a package will not reload if it is already loaded. With HPC, things get a little more complicated because there are multiple R processes ("workers") that each run one or more targets. A worker process that begins somewhere in the middle of the pipeline will have to load any required packages from scratch.

> Does the packages argument need to specify ALL packages required for the target, or only those in addition to those set globally in tar_option_set()?

`tar_option_set("packages")` is the default value of the `packages` argument in `tar_target()` (see https://docs.ropensci.org/targets/reference/tar_target.html#ref-usage) and any value you supply supersedes it. So if you supply any packages manually, you would need to supply all of them.
